jobbergate-api/jobbergate_api/s3_manager.py

```diff
@@ -1,37 +1,157 @@
     """
-    Provide a class for managing connection to an S3 bucket.
 
-    Prepares the bucket_name and s3 key template from project settings. Also initialize an s3
-    client. Provide methods as thin wrappers around s3 calls but provide the mappings for buckets and keys.
+    @abstractmethod
+    def put(self, file: typing.IO, key: str) -> None:
+        """
+        Upload a file to the client for the given key.
+        """
+        pass
+
+    @abstractmethod
+    def delete(self, key: str) -> None:
+        """
+        Delete a file from the client associated to the given key.
+        Raise KeyError if the file does not exist.
+        """
+        pass
+
+    @abstractmethod
+    def get(self, key: str) -> typing.IO:
+        """
+        Get a file from the client associated to the given key.
+        Raise KeyError if the file does not exist.
+        """
+        pass
+
+
+class DummyClient(BucketClientBase):
+    """
+    Dummy client designed to support the tests.
     """
 
     def __init__(self):
         """
-        Initialize the S3Manager class instance.
+        Initialize an dummy client.
+        """
+        self.client = {}
+
+    def put(self, file: typing.IO, key: str) -> None:
+        """
+        Upload a file to the client for the given key.
+        """
+        self.client[key] = file
+
+    def delete(self, key: str) -> None:
+        """
+        Delete a file from the client associated to the given key.
+        """
+        try:
+            del self.client[key]
+        except KeyError:
+            raise KeyError(f"No such key: {key}")
+
+    def get(self, key: str) -> typing.IO:
+        """
+        Get a file from the client associated to the given key.
+        """
+        try:
+            return self.client[key]
+        except KeyError:
+            raise KeyError(f"No such key: {key}")
+
+
+class S3Client(BucketClientBase):
+    """
+    S3 client.
+    """
+
+    def __init__(self):
+        """
+        Initialize an s3 client.
         """
         self.bucket_name = settings.S3_BUCKET_NAME
-        self.key_template = "applications/{app_id}/jobbergate.tar.gz"
         self.s3_client = boto3.client(
             "s3",
             endpoint_url=settings.S3_ENDPOINT_URL,
             aws_access_key_id=settings.AWS_ACCESS_KEY_ID,
             aws_secret_access_key=settings.AWS_SECRET_ACCESS_KEY,
         )
 
-    def _get_key(self, app_id):
+    def put(self, file: typing.IO, key: str) -> None:
+        """
+        Upload a file to the client for the given key.
+        """
+        self.s3_client.put_object(
+            Body=file,
+            Bucket=self.bucket_name,
+            Key=key,
+        )
+
+    def delete(self, key: str) -> None:
+        """
+        Delete a file from the client associated to the given key.
+        """
+        try:
+            self.s3_client.delete_object(Bucket=self.bucket_name, Key=key)
+        except self.s3_client.exceptions.NoSuchBucket:
```

Comment:
`NoSuchBucket` could be tested at initialization. Should we create a new bucket in that case?

Reply:
Fair point, but I've seen this error sometimes when testing locally. Is that a way to create the bucket from docker-compose?

Reply:
Exceptions for `NoSuchBucket` were removed in [75b4341](https://github.com/omnivector-solutions/jobbergate/pull/111/commits/75b434114eb0bbfac0917da9c87101bb9c9eae79)